src/Symfony/Component/Serializer/Normalizer/ChainDenormalizer.php

```diff
@@ -0,0 +1,242 @@
+            $genericType = class_exists($class) || interface_exists($class, false) ? 'object' : '*';
+
+            foreach ($this->denormalizers as $k => $denormalizer) {
+                if (!$denormalizer instanceof DenormalizerInterface) {
+                    continue;
+                }
+
+                $supportedTypes = $denormalizer->getSupportedTypes($format);
+
+                $doesClassRepresentCollection = str_ends_with($class, '[]');
+
+                foreach ($supportedTypes as $supportedType => $isCacheable) {
+                    if (\in_array($supportedType, ['*', 'object'], true)
+                        || $class !== $supportedType && ('object' !== $genericType || !is_subclass_of($class, $supportedType))
+                        && !($doesClassRepresentCollection && str_ends_with($supportedType, '[]') && is_subclass_of(strstr($class, '[]', true), strstr($supportedType, '[]', true)))
+                    ) {
+                        continue;
+                    }
+
+                    if (null === $isCacheable) {
+                        unset($supportedTypes['*'], $supportedTypes['object']);
+                    } elseif ($this->denormalizerCache[$format][$class][$k] = $isCacheable && $denormalizer->supportsDenormalization(null, $class, $format, $context)) {
+                        break 2;
+                    }
+
+                    break;
+                }
+
+                if (null === $isCacheable = $supportedTypes[\array_key_exists($genericType, $supportedTypes) ? $genericType : '*'] ?? null) {
+                    continue;
+                }
+
+                if ($this->denormalizerCache[$format][$class][$k] ??= $isCacheable && $denormalizer->supportsDenormalization(null, $class, $format, $context)) {
+                    break;
+                }
+            }
+        }
+
+        foreach ($this->denormalizerCache[$format][$class] as $k => $cached) {
+            $denormalizer = $this->denormalizers[$k];
+            if ($cached || $denormalizer->supportsDenormalization($data, $class, $format, $context)) {
+                return $denormalizer;
+            }
+        }
+
+        return null;
+    }
+
+    public function supportsDenormalization(mixed $data, string $type, string $format = null, array $context = []): bool
+    {
+        return isset(self::SCALAR_TYPES[$type]) || null !== $this->getDenormalizer($data, $type, $format, $context);
+    }
+
+    public function getSupportedTypes(?string $format): array
+    {
+        if (null === $format) {
+            $format = '__null__';
+        }
+
+        if (!isset($this->supportedCache[$format])) {
+            foreach ($this->denormalizers as $denormalizer) {
+                $this->supportedCache + $denormalizer->getSupportedTypes($format);
+            }
+        }
+
+        return $this->supportedCache[$format];
+    }
+
+    /**
+     * This method exists only for BC reasons. Will be removed in 8.0.
```

Comment:
Should it be tagged as internal?

Reply:
+ `@deprecated` to quickly spot it when it's time to